NY Times Most Viewed Article

sagar28pahwa opened the pull request:
## screenshots
![Simulator Screen Shot - iPhone 11 Pro - 2020-12-07 at 23 04 47](https://user-images.githubusercontent.com/22999737/101390046-3e140680-38e8-11eb-9917-5a5fc9f25d9c.png)
![Simulator Screen Shot - iPhone 11 Pro - 2020-12-07 at 23 04 00](https://user-images.githubusercontent.com/22999737/101390059-410ef700-38e8-11eb-9380-3db7a383efb8.png)
![Simulator Screen Shot - iPhone 11 Pro - 2020-12-07 at 23 04 12](https://user-images.githubusercontent.com/22999737/101390061-41a78d80-38e8-11eb-99d6-6f06015fb9af.png)
![Simulator Screen Shot - iPhone 11 Pro - 2020-12-07 at 23 20 14](https://user-images.githubusercontent.com/22999737/101390068-43715100-38e8-11eb-949d-483f997d0545.png)

## Code Coverage generated post UI test cases
<img width="803" alt="Screenshot 2020-12-07 at 11 59 22 PM" src="https://user-images.githubusercontent.com/22999737/101390288-977c3580-38e8-11eb-9ad4-8c29addce9a5.png">

## Code Coverage generated post unit test cases for viewModels:
1. NYTimesMostViewedArticleViewModel -> 100%
2. ArticleDetailViewModel -> 100%

serveshwar commented on NYTimesMostViewedArticles/NYTimesMostViewedArticleViewModel.swift:
```diff
@@ -0,0 +1,53 @@
+    }
+    
+    func showDetailScreen(index: Int, source: UIViewControllerType) {
+        if let articleDetailVC = UIStoryboard(name: "Main", bundle: nil).instantiateViewController(identifier: "ArticleDetailVC") as? ArticleDetailView {
```
Please declare all strings in a constant.

serveshwar replied:
Since this comment hasn't been implemented yet, hence please follow more better solution as explained below,

Since all the UI classes will be subclass of UIResponder hence instead of hardcoding reuseId in each class, we can create an extension of UIResponder and use it.

extension UIResponder { static var identifier: String { return "\(self)" } }

Instead of hardcoding "ArticleDetailVC" we can directly call ArticleDetailViewController.identifier` wherever required.
Please also follow this for ViewedArticleCell.

serveshwar commented on NYTimesMostViewedArticles/NetworkClient.swift:
```diff
@@ -0,0 +1,78 @@
+
+extension NetworkClientType {
+    func url(period: PeriodSection) -> URL {
+        let string = "\(API.baseURL)/\(API.contentURL)/\(period.rawValue).json?api-key=\(Keys.nyTimesKey.rawValue)"
```
Please use `URLComponent` instead of creating URL using hardcoded pattern.

serveshwar commented on NYTimesMostViewedArticles/NetworkClient.swift:
```diff
@@ -0,0 +1,78 @@
+extension NetworkClientType {
+    func url(period: PeriodSection) -> URL {
+        let string = "\(API.baseURL)/\(API.contentURL)/\(period.rawValue).json?api-key=\(Keys.nyTimesKey.rawValue)"
+        return URL(string: string)!
```
Force unwrapping should be avoided as it may lead to crash.

sagar28pahwa replied:
yes it is just that using optionals leads to littering of codebase with unwrapping whereas i know the url will exist since it the url format is respected in this case i.e it has a valid scheme, host, path.

serveshwar commented on NYTimesMostViewedArticles/NetworkClient.swift:
```diff
@@ -0,0 +1,78 @@
+    func url(period: PeriodSection) -> URL
+}
+
+enum Keys: String {
```
Please follow proper naming convention.

serveshwar commented on NYTimesMostViewedArticles/NetworkClient.swift:
```diff
@@ -0,0 +1,78 @@
+    }
+}
+
+class NetworkClient: NetworkClientType {
```
Currrent name `NetwokClient` is a verify generic name. Please use descriptive and contextual name here. As using a generic name will signify that all API calls being made from here this class will be heavily overloaded.

sagar28pahwa replied:
Yes it will be used for all api calls that why i have used generics in its implementation, in case of adding more variations of interaction with api like multi part we can refactor it, we have already services dedicated to specific apis, let me know if i misunderstood anything.

serveshwar commented on NYTimesMostViewedArticles/NetworkClient.swift:
```diff
@@ -0,0 +1,78 @@
+    func getMostViewedArticles(period: PeriodSection, completion: @escaping (Response?, Error?)->()) {
+
+        var request = URLRequest(url: self.url(period: period))
+        request.httpMethod = "GET"
```
1. All this general response parsing i.e. checking response status, mime, etc should not be part of each API method. Rather a root level network client should be written which will take care of all such things and will return json or error which respective network client functions will cast into required data model.

2. Also please try to use `Result` generic enum as defined here https://developer.apple.com/documentation/swift/result

serveshwar replied:
2nd point is still unresolved.

serveshwar commented on NYTimesMostViewedArticles/NetworkClient.swift:
```diff
@@ -0,0 +1,78 @@
+                return
+            }
+            guard let mime = response?.mimeType, mime == "application/json" else {
+                print("Wrong MIME type!")
```
Print command is only for debugging and should never be committed.

serveshwar commented on NYTimesMostViewedArticles/NetworkClient.swift:
```diff
@@ -0,0 +1,78 @@
+        request.httpMethod = "GET"
+        let task = session.dataTask(with: request, completionHandler: { data, response, error in
+            guard let data = data else {
+                return
```
There is no completion call back here. It will result in non-responsive UI.

serveshwar commented on NYTimesMostViewedArticles/NetworkClient.swift:
```diff
@@ -0,0 +1,78 @@
+            
+            guard let httpResponse = response as? HTTPURLResponse,
+                  (200...299).contains(httpResponse.statusCode) else {
+                return
```
There is no completion call back here. It will result in non-responsive UI.

serveshwar commented on NYTimesMostViewedArticlesUITests/NYTimesMostViewedArticlesUITests.swift:
```diff
@@ -0,0 +1,43 @@
+
+class NYTimesMostViewedArticlesUITests: XCTestCase {
+
+    override func setUpWithError() throws {
```
No UI tests written here.

serveshwar reviewed (commented):
Finished my review. I have added relevant comments where ever required. Below are some additional general points observed.
1. Please add app UI screenshots, unit test coverage report in MR description.
2. Better if you can also attach Sonarcube report here.
3. Unit Test coverage is too low. Please improve it.
4. MVVM approach should be followed properly. 
5. Use activity indicator for long running tasks like API calls, etc.
6. Proper project  folder hierarchy should be maintained.

sagar28pahwa commented:
> Finished my review. I have added relevant comments where ever required. Below are some additional general points observed.
> 
> 1. Please add app UI screenshots, unit test coverage report in MR description.
> 2. Better if you can also attach Sonarcube report here.
> 3. Unit Test coverage is too low. Please improve it.
> 4. MVVM approach should be followed properly.
> 5. Use activity indicator for long running tasks like API calls, etc.
> 6. Proper project  folder hierarchy should be maintained.

Hi @serveshwar please find the resolution/response to the above points
1. Added Screen Shots/ PR Description
2. Facing error while installation process of SonarQube, currently looking into it will update tomorrow
3. Improved Unit test cases Coverage
4. MVVM approach should be followed properly - please let me know if this is resolved as of now or not.
5. Activity Indicator incorporated for long running tasks like API calls, etc.
6. Proper project  folder hierarchy should be maintained. - please let me know if this is resolved as of now and if not please let me know the instances where it is violated

Thanks

serveshwar commented on NYTimesMostViewedArticles/NYTimesMostViewedArticleVC.swift:
```diff
@@ -0,0 +1,72 @@
+    
+    func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
+        tableView.deselectRow(at: indexPath, animated: true)
+        self.viewModel.showDetailScreen(index: indexPath.row, source: self)
```
Removing showDetail and directly push controller from here, doesn't solves MVVM problem.
Here, controller will call didSelect method on view model and viewModel should ask controller to showDetailScreen.

sagar28pahwa replied:
Please check new changes now

serveshwar commented on NYTimesMostViewedArticles/API/ArticleAPI.swift:
```diff
@@ -0,0 +1,36 @@
+    }
+}
+
+class ArticlesAPI: ArticlesAPIType {
```
This is more of a service class, please follow proper naming convention that should be self-descriptive.
Hence please rename it to MostPopularArticlesNetworkService

serveshwar commented on NYTimesMostViewedArticles/API/ArticleAPI.swift:
```diff
@@ -0,0 +1,36 @@
+    
+    func getMostViewedArticles(period: PeriodSection, completion: @escaping (Response?, Error?)->()) {
+        var request = URLRequest(url: ArticlesAPI.url(period: period))
+        request.httpMethod = "GET"
```
HttpMethod type should not be harcoded, rather its should be defined in a common enum and then that enum can be used here.

serveshwar commented:
Project is following more of MVC architecture rather then MVVM. View Model is just acting as data provider rather then decision taker.
Unit test coverage is good to be 90.2%

Most of the comments haven't been incorporated.

https://github.com/sagar28pahwa/NYTimesMostViewedArticles/pull/1#discussion_r526775887
https://github.com/sagar28pahwa/NYTimesMostViewedArticles/pull/1#discussion_r526790648
https://github.com/sagar28pahwa/NYTimesMostViewedArticles/pull/1#discussion_r526800798
https://github.com/sagar28pahwa/NYTimesMostViewedArticles/pull/1#discussion_r526748052
https://github.com/sagar28pahwa/NYTimesMostViewedArticles/pull/1#discussion_r526751270
https://github.com/sagar28pahwa/NYTimesMostViewedArticles/pull/1#discussion_r526761310
https://github.com/sagar28pahwa/NYTimesMostViewedArticles/pull/1#discussion_r526778011
https://github.com/sagar28pahwa/NYTimesMostViewedArticles/pull/1#discussion_r526781199
https://github.com/sagar28pahwa/NYTimesMostViewedArticles/pull/1#discussion_r526782400
https://github.com/sagar28pahwa/NYTimesMostViewedArticles/pull/1#discussion_r526786626
https://github.com/sagar28pahwa/NYTimesMostViewedArticles/pull/1#discussion_r526793213
https://github.com/sagar28pahwa/NYTimesMostViewedArticles/pull/1#discussion_r526795293
https://github.com/sagar28pahwa/NYTimesMostViewedArticles/pull/1#discussion_r526795499
https://github.com/sagar28pahwa/NYTimesMostViewedArticles/pull/1#discussion_r526797458
https://github.com/sagar28pahwa/NYTimesMostViewedArticles/pull/1#discussion_r526801602
https://github.com/sagar28pahwa/NYTimesMostViewedArticles/pull/1#discussion_r526803145
https://github.com/sagar28pahwa/NYTimesMostViewedArticles/pull/1#discussion_r526803159
https://github.com/sagar28pahwa/NYTimesMostViewedArticles/pull/1#discussion_r526805622
https://github.com/sagar28pahwa/NYTimesMostViewedArticles/pull/1#discussion_r526806741
https://github.com/sagar28pahwa/NYTimesMostViewedArticles/pull/1#discussion_r526807363
https://github.com/sagar28pahwa/NYTimesMostViewedArticles/pull/1#discussion_r526809280
https://github.com/sagar28pahwa/NYTimesMostViewedArticles/pull/1#discussion_r526810333
https://github.com/sagar28pahwa/NYTimesMostViewedArticles/pull/1#discussion_r526832641
https://github.com/sagar28pahwa/NYTimesMostViewedArticles/pull/1#discussion_r526832864
https://github.com/sagar28pahwa/NYTimesMostViewedArticles/pull/1#discussion_r526833097
https://github.com/sagar28pahwa/NYTimesMostViewedArticles/pull/1#discussion_r526836301
https://github.com/sagar28pahwa/NYTimesMostViewedArticles/pull/1#discussion_r526830668

New comments: 
https://github.com/sagar28pahwa/NYTimesMostViewedArticles/pull/1/files#r538102370
https://github.com/sagar28pahwa/NYTimesMostViewedArticles/pull/1/files#r538103671

serveshwar commented on NYTimesMostViewedArticles/ArticlesList/NYTimesMostViewedArticleViewController.swift:
```diff
@@ -0,0 +1,99 @@
+        tableView.register(UINib(nibName: cellNibName, bundle: Bundle.init(for: ViewedArticleCell.self)), forCellReuseIdentifier: cellNibName)
+        tableView.delegate = self
+        tableView.dataSource = self
+        tableView.isAccessibilityElement = true
```
All these cab be set from stroyboard itself.

sagar28pahwa replied:
set delegates and datasource in storyboard except couldnt find the accessibility option for tableView hence leaving it programmatically

serveshwar commented on NYTimesMostViewedArticles/CustomViews/ViewedArticleCell.swift:
```diff
@@ -0,0 +1,37 @@
+    }
+    
+    func configureAccessibility() {
+        titleLabel.isAccessibilityElement = true
```
All these cab be set from xib itself.

serveshwar commented on NYTimesMostViewedArticles/DetailScreen/ArticleDetailViewController.swift:
```diff
@@ -0,0 +1,43 @@
+        titleLabel.text = viewModel.title
+        source.text = viewModel.source
+        date.text = viewModel.date
+        titleLabel.accessibilityIdentifier = "titleLabel"
```
All these cab be set from stroyboard itself.

sagar28pahwa replied:
i tried setting accessibilityIdentifier from storyboard but it didn't work so didnt change it

serveshwar commented on NYTimesMostViewedArticles/DetailScreen/ArticleDetailViewController.swift:
```diff
@@ -0,0 +1,43 @@
+    }
+    
+    func configureAccessibility() {
+        titleLabel.isAccessibilityElement = true
```
All these cab be set from stroyboard itself.

serveshwar commented:
After 3rd (final) round of PR review below are the points those are pending,

Project is following more of MVC architecture rather then MVVM. View Model is just acting as data provider rather then decision taker.
Unit test coverage is good to be 90.2%

Most of the comments haven't been incorporated.

https://github.com/sagar28pahwa/NYTimesMostViewedArticles/pull/1#discussion_r526775887
https://github.com/sagar28pahwa/NYTimesMostViewedArticles/pull/1#discussion_r526790648
https://github.com/sagar28pahwa/NYTimesMostViewedArticles/pull/1#discussion_r526800798
https://github.com/sagar28pahwa/NYTimesMostViewedArticles/pull/1#discussion_r526748052
https://github.com/sagar28pahwa/NYTimesMostViewedArticles/pull/1#discussion_r526751270
https://github.com/sagar28pahwa/NYTimesMostViewedArticles/pull/1#discussion_r526761310
https://github.com/sagar28pahwa/NYTimesMostViewedArticles/pull/1#discussion_r526778011
https://github.com/sagar28pahwa/NYTimesMostViewedArticles/pull/1#discussion_r526781199
https://github.com/sagar28pahwa/NYTimesMostViewedArticles/pull/1#discussion_r526782400
https://github.com/sagar28pahwa/NYTimesMostViewedArticles/pull/1#discussion_r526786626
https://github.com/sagar28pahwa/NYTimesMostViewedArticles/pull/1#discussion_r526793213
https://github.com/sagar28pahwa/NYTimesMostViewedArticles/pull/1#discussion_r526795293
https://github.com/sagar28pahwa/NYTimesMostViewedArticles/pull/1#discussion_r526795499
https://github.com/sagar28pahwa/NYTimesMostViewedArticles/pull/1#discussion_r526797458
https://github.com/sagar28pahwa/NYTimesMostViewedArticles/pull/1#discussion_r526801602
https://github.com/sagar28pahwa/NYTimesMostViewedArticles/pull/1#discussion_r526803145
https://github.com/sagar28pahwa/NYTimesMostViewedArticles/pull/1#discussion_r526803159
https://github.com/sagar28pahwa/NYTimesMostViewedArticles/pull/1#discussion_r526805622
https://github.com/sagar28pahwa/NYTimesMostViewedArticles/pull/1#discussion_r526806741
https://github.com/sagar28pahwa/NYTimesMostViewedArticles/pull/1#discussion_r526807363
https://github.com/sagar28pahwa/NYTimesMostViewedArticles/pull/1#discussion_r526809280
https://github.com/sagar28pahwa/NYTimesMostViewedArticles/pull/1#discussion_r526810333
https://github.com/sagar28pahwa/NYTimesMostViewedArticles/pull/1#discussion_r526832641
https://github.com/sagar28pahwa/NYTimesMostViewedArticles/pull/1#discussion_r526832864
https://github.com/sagar28pahwa/NYTimesMostViewedArticles/pull/1#discussion_r526833097
https://github.com/sagar28pahwa/NYTimesMostViewedArticles/pull/1#discussion_r526836301
https://github.com/sagar28pahwa/NYTimesMostViewedArticles/pull/1#discussion_r526830668

New comments: 
https://github.com/sagar28pahwa/NYTimesMostViewedArticles/pull/1/files#r538102370
https://github.com/sagar28pahwa/NYTimesMostViewedArticles/pull/1/files#r538103671
https://github.com/sagar28pahwa/NYTimesMostViewedArticles/pull/1/files#r538115366
https://github.com/sagar28pahwa/NYTimesMostViewedArticles/pull/1/files#r538115460
https://github.com/sagar28pahwa/NYTimesMostViewedArticles/pull/1/files#r538115250
https://github.com/sagar28pahwa/NYTimesMostViewedArticles/pull/1/files#r538114843


We can merge this PR only when above points are resolved.

sagar28pahwa commented on NYTimesMostViewedArticles/Model/ViewedArticle.swift:
```diff
@@ -0,0 +1,110 @@
+
+public struct Response: Codable {
+
+    public var status: String?
```
Removed usage of Public

sagar28pahwa commented on NYTimesMostViewedArticles/NYTimesMostViewedArticleVC.swift:
```diff
@@ -0,0 +1,72 @@
+    }
+    
+    func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
+        tableView.deselectRow(at: indexPath, animated: true)
```
Using selection style as none would have been equivalent to using tableView.deselectRow(at: indexPath, animated: false) here i have set animated true it gives user indication that the selection had happen momentarily. please check the screenshot.
![Simulator Screen Shot - iPhone 11 Pro - 2020-12-08 at 14 04 25](https://user-images.githubusercontent.com/22999737/101459755-a4d40700-395e-11eb-9417-22346bd8b021.png)

sagar28pahwa commented on NYTimesMostViewedArticles/NYTimesMostViewedArticleViewModel.swift:
```diff
@@ -0,0 +1,53 @@
+    }
+    
+    func fetchMostViewArticles(completion: @escaping (Error?)->()) {
+        self.networkClient.getMostViewedArticles(period: .day) { (response, error) in
```
added default arguement so that it is configurable

sagar28pahwa commented on NYTimesMostViewedArticles/NYTimesMostViewedArticleVC.swift:
```diff
@@ -0,0 +1,72 @@
+}
+
+class NYTimesMostViewedArticleVC: UIViewController {
+    @IBOutlet weak var tableView: UITableView!
```
marked @IBOutlet as private and not just private (set)

sagar28pahwa commented on NYTimesMostViewedArticles/NYTimesMostViewedArticleViewModel.swift:
```diff
@@ -0,0 +1,53 @@
+            else if let error = error {
+                completion(error)
+            }
+        }
```
although the query is reasonable please note that one of them will exists has been ensured in the NetworkClient class since there is no case there where both of them can be nil

sagar28pahwa commented on NYTimesMostViewedArticles/DetailScreen/ArticleDetailVC.swift:
```diff
@@ -0,0 +1,32 @@
+    
+    @IBOutlet weak var date: UILabel!
+    
+    var viewModel: ArticleDetailViewModel!
```
agree

sagar28pahwa commented on NYTimesMostViewedArticles/NYTimesMostViewedArticleVC.swift:
```diff
@@ -0,0 +1,72 @@
+        self.tableView.dataSource = self
+    }
+    
+    func fetchMostViewedArticles() {
```
@serveshwar Please check the latest changes.
